[Pytorch] Only print forward gradient of a tensor if autograd is built

JacobSzwejbka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54446 [Pytorch] Only print forward gradient of a tensor if autograd is built**

title

Differential Revision: [D27239164](https://our.internmc.facebook.com/intern/diff/D27239164/)

Commit:
[Pytorch] Only print gradient of a tensor if it requires_grad

title

Differential Revision: [D27239164](https://our.internmc.facebook.com/intern/diff/D27239164/)

[ghstack-poisoned]

albanD reviewed (changes requested):
This doesn't look right.
The requires_grad field is for backward AD and the code you change here is for forward AD.

What is the problem that this tries to solve?

JacobSzwejbka commented:
> This doesn't look right.
> The requires_grad field is for backward AD and the code you change here is for forward AD.
> 
> What is the problem that this tries to solve?

Several users on mobile have run into a problem where autograd isnt loaded causing a crash when the attempt to print the results of calling forward. This change mitigates that issue somewhat (since these instances are eval() models trying to just see the output). A more robust solution would be to dynamically check if autograd is built before trying to print the gradient, but I thought something like this could work in the meantime since the gradient isnt of any interest here. Would I ever want to print the gradient if requires_grad=False? Whats it even mean to do that?

albanD commented:
Forward and backward mode are completely independent subsystem in the autograd. They allow to compute gradient-like quantities with very different approaches and so depending on the user's function, they use one or the other.
The `.requires_grad` field is only relevant for backward mode.

How is the requires_grad() acessor overwritten when autograd is not builtin?
I guess we want to do the same for the `_fw_grad()` method to always return an undefined Tensor when autograd is not compiled in.

JacobSzwejbka commented:
> Forward and backward mode are completely independent subsystem in the autograd. They allow to compute gradient-like quantities with very different approaches and so depending on the user's function, they use one or the other.
> The `.requires_grad` field is only relevant for backward mode.
> 
> How is the requires_grad() acessor overwritten when autograd is not builtin?
> I guess we want to do the same for the `_fw_grad()` method to always return an undefined Tensor when autograd is not compiled in.

I'll look into whats going on with requires_grad() when autograd is not built and change the diff to try and copy that in fw_grad

JacobSzwejbka commented:
> Forward and backward mode are completely independent subsystem in the autograd. They allow to compute gradient-like quantities with very different approaches and so depending on the user's function, they use one or the other.
> The `.requires_grad` field is only relevant for backward mode.
> 
> How is the requires_grad() acessor overwritten when autograd is not builtin?
> I guess we want to do the same for the `_fw_grad()` method to always return an undefined Tensor when autograd is not compiled in.

Ok I looked into it. "requires_grad()" looks if autograd_meta_ is defined, and if its not it just returns false. This works out fine if autograd isnt built into the app because autograde_meta_ wont be defined and we will just return false. _fw_grad has the same logic for checking autograd_meta_ but it returns a constant tensor reference so it returns impl::GetAutogradMetaFactory()->undefined_tensor(); and GetAutogradMetaFactory() is the call that requires autograd to exist.

albanD commented:
Ho ok,
I guess we will need to add a "has_fw_grad" that returns a bool and that we can use to guard such code. Will that work for you?

Commit:
Update on "[Pytorch] Only print gradient of a tensor if it requires_grad"


title

Differential Revision: [D27239164](https://our.internmc.facebook.com/intern/diff/D27239164/)

[ghstack-poisoned]

albanD commented on aten/src/ATen/core/Formatting.cpp:
```diff
@@ -295,14 +295,13 @@ std::ostream& print(std::ostream& stream, const Tensor & tensor_, int64_t linesi
       if (fw_grad.defined()) {
         stream << ", tangent:" << std::endl << fw_grad;
       }
+      stream << " ]";
```
That one should be outside. It closes an opening bracket that is always present.

albanD commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -710,6 +710,14 @@ class TORCH_API Tensor {
     impl_->_set_fw_grad(new_grad, *this, level, is_inplace_op);
   }
 
+  /// This function returns if the forward gradient for this tensor exists at any level
```
"return true"?

albanD commented on c10/core/TensorImpl.cpp:
```diff
@@ -51,6 +51,13 @@ const at::Tensor& TensorImpl::_fw_grad(uint64_t level, const at::Tensor& self) c
+  // One reason for autograd_meta_ to be undefined is autograd isnt enabled.
+  // However in functions like _fw_grad the fallback code path also requires
+  // autograd to be enabled. This function can serve as a guard then.
+  return autograd_meta_ != nullptr;
```
This will return True even if there is no actual fw grad.
If you want a function that just checks for the existence of autograd meta, you can just use `autograd_meta()` and check if it is null.

albanD reviewed (changes requested):
Thanks for the update.
I think that you actually just want to add a `if(tensor.autograd_meta())` to protect the code in Formatting.cpp. That should guard things properly for you. without adding the has_fw_grad function that would be harder to implement as you will actually have to call into the autograd code to check if any forward grad exists.

Commit:
Update on "[Pytorch] Only print forward gradient of a tensor if autograd is built"


title

Differential Revision: [D27239164](https://our.internmc.facebook.com/intern/diff/D27239164/)

[ghstack-poisoned]

JacobSzwejbka commented:
Ok I think I've arrived at what I wanted. Use the existence of autograd_meta_ as a proxy for if autograd exists. If it doesn't just skip the print for the forward gradient otherwise the existing behavior is fine. Thanks for all the pointers Alban!

JacobSzwejbka commented on aten/src/ATen/core/Formatting.cpp:
```diff
@@ -295,13 +295,15 @@ std::ostream& print(std::ostream& stream, const Tensor & tensor_, int64_t linesi
+    // Proxy check for if autograd was built
+    if (tensor.getIntrusivePtr()->autograd_meta()) {
       auto& fw_grad = tensor._fw_grad(/* level */ 0);
       if (fw_grad.defined()) {
```
Is this check still necessary?

albanD replied:
Yes because we don't want to print all the undefined Tensors you get for things that don't have a forward grad.

albanD reviewed (approved):
LGTM, thanks for taking the time to update this!
